### PR TITLE
Add implicit convertibility to function pointer with higher state mutability

### LINKED ISSUE #4349: Make pure function pointers implicitly convertible to view function pointers
Solidity version: 0.4.24

If the parameter types and return value types match, a pure function pointer should be implicitly convertible to a view function pointer. The same goes for implicit convertibility from view to effectful, and from pure to effectful. It's the same principle as an implicit cast from `uint128` to `uint256`: going from less possibilities to a superset of possibilities.  For example:

```
pragma solidity ^0.4.24;
contract Test
{
    function func1(uint256 x) internal pure returns (uint256)
    {
        return x;
    }
    function func2(function(uint256)view returns(uint256) a) internal view returns (uint256)
    {
        return a(4);
    }
    function func3() internal view returns (uint256)
    {
        return func2(func1);
    }
}
```

This currently gives a TypeError at the `func2(func1)`:

> TypeError: Invalid type for argument in function call. Invalid implicit conversion from function (uint256) pure returns (uint256) to function (uint256) view returns (uint256) requested.

I would like so get around this problem, so is there a way to explicitly convert the function pointer type?

Otherwise it seems like I will have to either downgrade all my pure functions to view or use inline assembly.

### PR DESCRIPTION
**Pull request by JesseBusman**
Fixes #4349

### CODE REVIEW

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -2607,6 +2607,41 @@ bool FunctionType::isExplicitlyConvertibleTo(Type const& _convertTo) const
...
+	if (
+		m_kind != convertTo.m_kind ||
+		m_parameterTypes.size() != convertTo.m_parameterTypes.size() ||
+		m_returnParameterTypes.size() != convertTo.m_returnParameterTypes.size()
+	)
+		return false;
+	
+	auto typeCompare = [](TypePointer const& _a, TypePointer const& _b) -> bool { return *_a == *_b; };
+	if (
+		!equal(m_parameterTypes.cbegin(), m_parameterTypes.cend(), convertTo.m_parameterTypes.cbegin(), typeCompare) ||
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: There is a function `hasEqualArgumentTypes` - you could add another called `hasEqualArgumentAndReturnTypes` and call it here. Actually in that function it should be `Parameters`, not `Arguments` - could you also fix that at the same time, please?

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Ah now I see that most of this is taken from `operator==` - could you reduce code duplication somehow?

File: docs/types.rst [hunk 2]
```diff
@@ -389,6 +389,13 @@ There are two ways to access a function in the current contract: Either directly
 by its name, ``f``, or using ``this.f``. The former will result in an internal
 function, the latter in an external function.
 
+A function type A is implicitly convertible to a function type B if and only if
+their parameter types are identical, their return types are identical,
+their internal/external property is identical and the state mutability of A
```
Comment (chriseth) [inline docs/types.rst]: Perhaps replace `smaller than` by `less restrictive`?

**Review by chriseth — CHANGES_REQUESTED** [review]
Please add syntax tests which do implicit conversions between "actual" function types (i.e. pointers) and also from things like `this.f` or `f` if `f` is the name of an actual function in the contract.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -2607,6 +2607,41 @@ bool FunctionType::isExplicitlyConvertibleTo(Type const& _convertTo) const
...
+		!equal(m_returnParameterTypes.cbegin(), m_returnParameterTypes.cend(), convertTo.m_returnParameterTypes.cbegin(), typeCompare)
+	)
+		return false;
+	
+	if (m_gasSet != convertTo.m_gasSet || m_valueSet != convertTo.m_valueSet)
+		return false;
+	if (bound() != convertTo.bound())
+		return false;
+	if (bound() && *selfType() != *convertTo.selfType())
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: The check for `selfType` should actually be included in the parameterTypes check, but we should at least have an assertion here about that.

File: libsolidity/ast/Types.cpp [hunk 4]
```diff
@@ -2607,6 +2607,41 @@ bool FunctionType::isExplicitlyConvertibleTo(Type const& _convertTo) const
...
+		return false;
+
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Where do you check for internal / external?

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Ah sorry, that's `m_kind`.

**Comment by randomnetcat** [comment]
Suggestion for code duplication: have operator== check implicitly convertible and then check anything else that can vary implicitly (state mutability).

Also, I think the conversion needs to be different than just allow a less strict state mutability. It should not be permissible to implicitly convertible any state mutability that is not payable to payable, since if the taget type is payable but the function is not, sending ether would throw, which seems to violate payable guarantee.

Also, perhaps a payable function should be able to be stored in a nonpayable pointer, since no other state guarantees are made, and it is permissible not to send ether to a payable function, so no guarantees are violated.

**Comment by JesseBusman** [comment]
@meowingtwurtle Yeah, I was just implementing operator== in terms of isImplicitlyConvertible :)

Good point about the payable stuff.

File: docs/types.rst (hunk 2 above)
Comment (JesseBusman) [reply docs/types.rst]: Done

File: libsolidity/ast/Types.cpp (hunk 1 above)
Comment (JesseBusman) [reply libsolidity/ast/Types.cpp]: I've reduced code duplication by implementing operator== in terms of isImplicitlyConvertible

File: libsolidity/ast/Types.cpp (hunk 3 above)
Comment (JesseBusman) [reply libsolidity/ast/Types.cpp]: I've turned it into an assertion

File: test/libsolidity/syntaxTests/types/function_parameter_return_types.sol [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    	function (string memory) internal pure returns (string memory) var_string_to_string = string_to_string;
+    	
+    	function (uint256, uint256) internal pure returns (uint256) var_uint256_uint256_to_uint256 = uint256_uint256_to_uint256;
+    	function (uint256, uint256) internal pure returns (string memory) var_uint256_uint256_to_string = uint256_uint256_to_string;
+    	function (string memory, uint256) internal pure returns (string memory) var_string_uint256_to_string = string_uint256_to_string;
+    	function (string memory, string memory) internal pure returns (string memory) var_string_string_to_string = string_string_to_string;
+    	function (uint256, string memory) internal pure returns (string memory) var_uint256_string_to_string = uint256_string_to_string;
+    	
+    	// Should fail:
+    	var_uint256_to_uint256 = uint256_to_string;
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/types/function_parameter_return_types.sol]: Please split succeeding and failing conversions into two different tests, since nobody cares about the source locations below :)

File: libsolidity/ast/Types.cpp [hunk 6]
```diff
@@ -2570,28 +2570,10 @@ bool FunctionType::operator==(Type const& _other) const
...
-	if (
-		!equal(m_parameterTypes.cbegin(), m_parameterTypes.cend(), other.m_parameterTypes.cbegin(), typeCompare) ||
-		!equal(m_returnParameterTypes.cbegin(), m_returnParameterTypes.cend(), other.m_returnParameterTypes.cbegin(), typeCompare)
-	)
-		return false;
-	//@todo this is ugly, but cannot be prevented right now
-	if (m_gasSet != other.m_gasSet || m_valueSet != other.m_valueSet)
-		return false;
-	if (bound() != other.bound())
+	if (!isImplicitlyConvertibleTo(other))
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: This reads a little weird (is it really symmetric?). You might want to extract the common parts into a function whose name is a little more harmless (`equalUpToStateMutability`).

File: test/libsolidity/syntaxTests/types/function_state_mutability.sol [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    	funcTakesExternalPayable(this.externalMutableFunc); // mutable -> payable should fail
+    	
+    	funcTakesExternalMutable(this.externalPureFunc);
+    	funcTakesExternalMutable(this.externalViewFunc);
+    	funcTakesExternalMutable(this.externalPayableFunc);
+    	funcTakesExternalMutable(this.externalMutableFunc);
+    }
+}
+// ----
+// TypeError: (1675-1691): Invalid type for argument in function call. Invalid implicit conversion from function (uint256) view returns (uint256) to function (uint256) pure returns (uint256) requested.
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/types/function_state_mutability.sol]: Same here, it's not really visible what fails and what succeeds.

**Comment by chriseth** [comment]
@JesseBusman thanks for your contribution! Do you still have time to finish this? If not, please tell us so someone else can take over and do the last missing bits.

**Comment by JesseBusman** [comment]
@chriseth It's vacation time, but I think I can do it this evening.

File: docs/types.rst [hunk 8]
```diff
@@ -389,6 +389,14 @@ There are two ways to access a function in the current contract: Either directly
 by its name, ``f``, or using ``this.f``. The former will result in an internal
 function, the latter in an external function.
 
+A function type A is implicitly convertible to a function type B if and only if
```
Comment (axic) [inline docs/types.rst]: While not necessary, I think consistently marking `A` and `B` with a different font (italic or code highlight aka backticks) could help readability.

File: libsolidity/ast/Types.cpp [hunk 9]
```diff
@@ -2852,7 +2870,7 @@ bool FunctionType::canTakeArguments(TypePointers const& _argumentTypes, TypePoin
 		);
 }
 
-bool FunctionType::hasEqualArgumentTypes(FunctionType const& _other) const
+bool FunctionType::hasEqualParameterTypes(FunctionType const& _other) const
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: Can you pull this change into a separate PR? Look like a good change we could merge easily.

**Comment by axic** [comment]
Needs to be rebased and also lacks a changelog entry.

**Comment by JesseBusman** [comment]
@axic Can some experienced git user do this? I have no idea how to rebase, or split a PR into two.

**Comment by meowingtwurtle** [comment]
If you don't want to figure it out, make sure the box that says allow
mainteners to edit PR is checked.

To rebase:

`git fetch origin/develop`
`git rebase origin develop`
(Just exit the editor that opens)
Fix any conflicts, then `git add` them, then `git rebase --continue`
`git push -f`

If you need to split a PR into two, you need to remove the change from this
PR. Rebase away the commit (`git rebase origin/develop` then change the
line with the commit from pick to drop). If you need to fix files in other
commits that depended on the change, change the word on that line to edit,
then make changes when you reach it, followed by `git add` and `git rebase
--continue`

**Comment by randomnetcat** [comment]
If you don't want to figure it out, make sure the box that says allow
mainteners to edit PR is checked.

To rebase:

```git fetch origin/develop```
```git rebase origin develop```
(Just exit the editor that opens)
Fix any conflicts, then ```git add``` them, then ```git rebase --continue```
```git push -f```

If you need to split a PR into two, you need to remove the change from this
PR. Rebase away the commit (```git rebase origin/develop``` then change the
line with the commit from pick to drop). If you need to fix files in other
commits that depended on the change, change the word on that line to edit,
then make changes when you reach it, followed by ```git add``` and ```git rebase
--continue```

**Comment by chriseth** [comment]
I'm taking this over.

File: libsolidity/ast/Types.h [hunk 10]
```diff
@@ -1032,7 +1033,11 @@ class FunctionType: public Type
 	/// expression the function is called on.
 	bool canTakeArguments(TypePointers const& _arguments, TypePointer const& _selfType = TypePointer()) const;
 	/// @returns true if the types of parameters are equal (does't check return parameter types)
-	bool hasEqualArgumentTypes(FunctionType const& _other) const;
+	bool hasEqualParameterTypes(FunctionType const& _other) const;
+	/// @returns true iff the return types are equal (does't check parameter types)
+	bool hasEqualReturnTypes(FunctionType const& _other) const;
+	/// @returns true iff the parameter types and the return types are equal
+	bool hasEqualParameterAndReturnTypes(FunctionType const& _other) const;
```
Comment (axic) [inline libsolidity/ast/Types.h]: I think these could also be pulled out and even used by the "function shadowing" type checker feature.

**Comment by chriseth** [comment]
@axic please review - will squash later.

**Review by chriseth — DISMISSED** [review]

File: libsolidity/ast/Types.cpp [hunk 11]
```diff
@@ -2543,28 +2543,10 @@ bool FunctionType::operator==(Type const& _other) const
...
-	)
-		return false;
-	//@todo this is ugly, but cannot be prevented right now
-	if (m_gasSet != other.m_gasSet || m_valueSet != other.m_valueSet)
-		return false;
-	if (bound() != other.bound())
+	if (!equalUpToStateMutability(other))
 		return false;
-	if (bound() && *selfType() != *other.selfType())
+	if (m_stateMutability != other.stateMutability())
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: ~~If this is here, what is the point of `!equalUpToStateMutability` ?~~

Right, the name wasn't clicking at the first second.

Comment (axic) [reply libsolidity/ast/Types.cpp]: Maybe `equalExcludingStateMutability` ?

File: test/libsolidity/syntaxTests/types/function_types/function_parameter_return_types_fail.sol [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    
+    function uint256_uint256_to_uint256(uint256 x, uint256 y) internal pure returns (uint256) { return x + y; }
+    function uint256_uint256_to_string(uint256 x, uint256 y) internal pure returns (string memory) { return x == y ? "a" : "b"; }
+    function string_uint256_to_string(string memory x, uint256 y) internal pure returns (string memory) { return y == 0 ? "a" : x; }
+    function string_string_to_string(string memory x, string memory y) internal pure returns (string memory) { return bytes(x).length == 0 ? y : x; }
+    function uint256_string_to_string(uint256 x, string memory y) internal pure returns (string memory) { return x == 0 ? "a" : y; }
+    
+    function tests() internal pure
+    {
+    	// Should succeed:
```
Comment (axic) [inline test/libsolidity/syntaxTests/types/function_types/function_parameter_return_types_fail.sol]: I think we had the rule to split succeeding and failing tests.

File: libsolidity/ast/Types.cpp [hunk 13]
```diff
@@ -2837,6 +2844,38 @@ bool FunctionType::hasEqualParameterTypes(FunctionType const& _other) const
...
+		return false;
+
+	//@todo this is ugly, but cannot be prevented right now
+	if (m_gasSet != _other.m_gasSet || m_valueSet != _other.m_valueSet)
+		return false;
+
+	if (bound() != _other.bound())
+		return false;
+
+	solAssert(!bound() || *selfType() == *_other.selfType(), "");
```
Comment (axic) [inline libsolidity/ast/Types.cpp:2900]: Why is this an assert? Because we cannot express it in the type system (in a test)?

File: libsolidity/ast/Types.cpp [hunk 14]
```diff
@@ -2580,6 +2562,31 @@ bool FunctionType::isExplicitlyConvertibleTo(Type const& _convertTo) const
...
+
+	if (!equalUpToStateMutability(convertTo))
+		return false;
+
+	// non-payable should not be convertible to payable
+	if (m_stateMutability != StateMutability::Payable && convertTo.stateMutability() == StateMutability::Payable)
+		return false;
+	
+	// payable should be convertible to non-payable, because you are free to pay 0 ether
+	if (m_stateMutability == StateMutability::Payable && convertTo.stateMutability() == StateMutability::NonPayable)
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: Can this be misused for putting assigning a non-payable function to where payable is expected?

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Can you provide an example? This should allow assignments of payable functions to non-payable functions, but not vice-versa.

File: libsolidity/ast/Types.cpp (hunk 13 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:2900]: This checks that `hasEqualParameterTypes` also checks the bound type if the function is bound.

File: test/libsolidity/syntaxTests/types/function_types/function_parameter_return_types_fail.sol (hunk 12 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/types/function_types/function_parameter_return_types_fail.sol]: I split them but forgot to remove the comment - these actually all fail.

**Comment by axic** [comment]
This is failing tests now.

Will do another review when they are running again.

**Comment by chriseth** [comment]
Updated.

File: docs/types.rst [hunk 15]
```diff
@@ -413,8 +413,18 @@ function type should not return anything, the whole ``returns (<return types>)``
 part has to be omitted.
 
 By default, function types are internal, so the ``internal`` keyword can be
-omitted. In contrast, contract functions themselves are public by default,
-only when used as the name of a type, the default is internal.
+omitted. In contrast, visibility has to be specified for contract functions themselves,
+only when used as a type, there is a default of internal.
```
Comment (ekpyron) [inline docs/types.rst]: Maybe a full stop before this line would make it easier to read? Also why ``themselves``?
```
In contrast, visibility has to be specified for all contract functions.
For function types the visibility defaults to internal.
```
By the way: should we actually remove the exception for function types here? Might be more consistent.

File: docs/types.rst [hunk 16]
```diff
@@ -413,8 +413,18 @@ function type should not return anything, the whole ``returns (<return types>)``
...
+
+A function type ``A`` is implicitly convertible to a function type ``B`` if and only if
+their parameter types are identical, their return types are identical,
+their internal/external property is identical and the state mutability of ``A``
```
Comment (ekpyron) [inline docs/types.rst]: Maybe ``visibility`` instead of ``internal/external property``?

File: docs/types.rst [hunk 17]
```diff
@@ -413,8 +413,18 @@ function type should not return anything, the whole ``returns (<return types>)``
...
+is not more restrictive than the state mutability of ``B``. In this sense,
+``view`` functions are considered most restrictive and ``payable`` least
```
Comment (ekpyron) [inline docs/types.rst]: This should probably be ``pure`` not ``view``.
~Maybe we should just mention the order ``pure > view > payable > nonpayable``.~

**Review by ekpyron — COMMENTED** [review]
Apart from the comments about the documentation, this looks good to me!

File: docs/types.rst (hunk 17 above)
Comment (ekpyron) [reply docs/types.rst]: Hm - the order is actually ``pure > view > nonpayable`` and ``payable > nonpayable``, but not ``view > payable``.

File: docs/types.rst (hunk 16 above)
Comment (chriseth) [reply docs/types.rst]: Hm, this is not really about visibility. For function definitions, `internal/external` can be derived from the visibility, but for function types this is more specific and "visibilty" would be the wrong term.

Comment (chriseth) [reply docs/types.rst]: But yeah, we should probably find a term for this at some point...

Comment (ekpyron) [reply docs/types.rst]: True.

File: docs/types.rst (hunk 17 above)
Comment (chriseth) [reply docs/types.rst]: Right, non-payable is least restrictive.

Comment (ekpyron) [reply docs/types.rst]: Maybe we should divide this into ``and the state mutability of A is not more restrictive than the state mutability of B, except if one of them is payable. A payable function type can be converted to a nonpayable function type, but no type that is not also payable can be converted to a payable function type.``

...or something better - in any case I would make the entire order explicit here.

Comment (ekpyron) [reply docs/types.rst]: Maybe we should even put ``you are free to pay zero ether`` and ``sending ether to non-payable functions is not allowed`` in parenthesis here.

File: docs/types.rst [hunk 18]
```diff
@@ -426,8 +426,25 @@ function type should not return anything, the whole ``returns (<return types>)``
...
-only when used as the name of a type, the default is internal.
+omitted. Note that this only applies to function types. Visibility has
+to be specified explicitly for functions defined in contracts, they
+do not have a default.
+
+A function type ``A`` is implicitly convertible to a function type ``B`` if and only if
+their parameter types are identical, their return types are identical,
+their internal/external property is identical and the state mutability of ``A``
+is not more restrictive than the state mutability of ``B``. In particular:
+``pure`` functions can be converted to ``view`` and ``non-payable`` functions and
```
Comment (ekpyron) [inline docs/types.rst]: I had to read this three times to correctly understand the implicit parenthesis in the sentence. We should use full stops instead:
```
In particular: ``pure`` functions can be converted to ``view`` and ``non-payable`` functions.
``view`` functions can be converted to ``non-payable`` functions.
``payable`` functions can be converted to ``non-payable`` functions.
```

Apart from that I'm fine with it now.

Comment (chriseth) [reply docs/types.rst]: I will even make a list out of it!

Comment (chriseth) [reply docs/types.rst]: Also removed the sentence about what is least restrictive, it is just confusing.

**Review by ekpyron — DISMISSED** [review]
